File: typescript/parser/src/parser/common/option-parser.ts
import { Parser } from './parser';
import { ParseResult, ParseSuccess, ParseFailer } from './parser-result';
import { Option, Some, None } from './type/option'

export class OptionParser<T> implements Parser<Option<T>> {
	parser: Parser<T>
	constructor(parser: Parser<T>) {
		this.parser = parser
	}

	parse(input: string): ParseResult<Option<T>> {
		const result = this.parser.parse(input)
		if(result instanceof ParseSuccess) {
			return new ParseSuccess(new Some(result.value), result.next)
		} else {
			return new ParseSuccess(new None(), result.next)
		}
	}
}

File: java/parsers_java8/src/main/java/jp/co/teruuu/parser/common/OptionParser.java
package jp.co.teruuu.parser.common;

import java.util.Optional;

public class OptionParser<T> implements Parser<Optional<T>> {
    private Parser<T> parser;
    public OptionParser(Parser<T> parser) {
        this.parser = parser;
    }

    @Override
    public ParseResult<Optional<T>> parse(String input) {
        ParseResult<T> result = parser.parse(input);
        if(result instanceof ParseResult.Failure<?>) {
            return new ParseResult.Success<>(Optional.ofNullable(null), input);
        } else {
            ParseResult.Success<T> success = (ParseResult.Success<T>)result;
            return new ParseResult.Success<>(Optional.of(success.value), success.next);
        }
    }
}